### PR TITLE
Move built runtime and compiler objects to top-level build directory

### PR DESCRIPTION
**Pull request by mppf**
This PR changes the runtime and compiler Makefiles to store built objects in a top-level build directory. This directory is removed on 'make cleanall' or 'make clobber'. This addresses problems where some .o files were not removed on make cleanall or make clobber.

The directory hierarchy is like this:

```
  $CHPL_HOME/build/
        compiler/
            $CHPL_MAKE_RUNTIME_SUBDIR (configuration dependent)
        runtime/
            $CHPL_MAKE_RUNTIME_SUBDIR (configuration dependent)
        runtime-launcher/
            $CHPL_MAKE_LAUNCHER_SUBDIR (configuration dependent)
```

Additionally, remove TAGS or tags on make clean.

TODO:

  make clean should remove SysCTypes.chpl.

Testing:

 - [x] make check on darwin with quickstart
 - [x] make check on darwin with quickstart+GASNet
 - [x] make check on darwin with default settings
 - [x] make check on linux64 with quickstart
 - [x] make check on linux64 with default settings+GASNet
 - [x] make check on linux64 with default settings+GASNet+IBV+slurm-gasnetrun_ibv

Reviewed by @sungeunchoi - thanks!

### CODE REVIEW

**Comment by sungeunchoi** [comment]
This only updates clobber the runtime directory.  You'll want the same behavior in the compiler directory.

Since you're here, I think these should be addressed, tho not necessarily in this PR:
- `make clean` does not remove the appropriate SysCTypes.chpl 
- `make clobber` should remove compiler/TAGS (it does so in runtime and modules)
- `make clobber` in the test directory should also clean (i.e., run start_test --clean-only)

**Comment by sungeunchoi** [comment]
👍  Please add the above to your issue tracker or TODO list.

**Comment by bradcray** [comment]
Taking a quick look, it strikes me as being more fragile to require subdirectories to be explicitly listed in Makefiles rather than to lean on a 'find' command.  It also raises questions about whether 'muxed' and 'ugni' should be listed here.  What downsides would using a 'find' lead to?

**Comment by mppf** [comment]
re. find, it seems more "safe" and "predictable" to use the directory listing approach.
Of course, I could adjust the Makefiles to compute ALL_SUBDIRS as all directories with sub-makefiles. But, that adds more scripts/weird complexity that I'd rather avoid.

Anyway, I don't think it's a particular maintenance burden to list the directories in the Makefiles. It's pretty typical to do so, actually... although in these particular Makefiles we've been just building the 1 configured option so didn't.

**Comment by mppf** [comment]
Oh yeah, the other thing we could do - which might be better - would be to store all of these 'gen' directories in $CHPL_HOME/build in a way that mirrors the source directory hierarchy. Then we could just blow away $CHPL_HOME/build on a make cleanall / make clobber.

I think that would be the nicest of all, but it's a different sort of change.

**Comment by sungeunchoi** [comment]
A find-based solution always makes me a bit wary, more so with the generic name for the gen directory (which could be changed I suppose).  Long term, I'm in favor in general of a separate build directory.

**Comment by bradcray** [comment]
I like the separate build directory idea too if it doesn't preclude our ability to have multiple configurations built simultaneously (which it seems like it shouldn't).  In the current scheme, what's the plan for 'ungi' and 'muxed'?

**Comment by mppf** [comment]
I don't think the scheme in this PR has a good strategy for ugni and muxed. We could keep going with this PR now, but my preference would be to try to make the separate build directory. (That covers ugni and muxed nicely).

**Comment by sungeunchoi** [comment]
A few questions for the build dir solution:

- Should documentation move to the build dir?
- Should the TAGS file move to the build dir?
- If the TAGS file did move, perhaps it should be specific to that particular build (i.e., only include the relevant sources for that build)?

**Comment by mppf** [comment]
Anybody know why we're not removing the .d files on a 'make clean' ?

**Comment by bradcray** [comment]
mppf: The lack of removal of .d's could date back from the time when dependences had to be computed manually as a separate step and there was no point in paying that cost again.

I feel mixed about Sung's proposal to move the TAGS files into the build dirs.  On one hand, it might be nice to have multiple reasonable TAGS files available at any given time.  On the other hand, it seems more painful to have to dig into long unique-ified directory names in order to find the right one.  I.e., it's nice that compiler/TAGS is the compiler tags and runtime/TAGS is the runtime's tags.  A compromise might be to have the top-level TAGS be a symbolic link to the most-recently built configuration's TAGS file.

I'm gonna let those with a stronger vision for the docs directory (like @ben-albrecht) weigh in on the docs question.

**Comment by ben-albrecht** [comment]
> Should documentation move to the build dir?

Thinking about this in terms of post-#4710 merge...

There are 2 kinds of generated documentation: 

* the generated rst files build into various places among the static rst files all under `doc/`
* the generated html via sphinx built into `doc/build`

Placing the generated rst files in a top-level `build/` would impose some challenges and wouldn't make much sense.

The generated HTML could be built into the top-level build directory if it provides any additional benefit. In #4710, it's built into `doc/build/`, which is wiped with a `make cleanall`

**Comment by mppf** [comment]
Updated. This PR now:

* uses a top-level build directory
* removes TAGS files on make clean

I believe that the remaining TODO in this area is for 'make clean' to remove SysCTypes.chpl.

**Comment by sungeunchoi** [comment]
Just to confirm, by *transitionally* above in your PR description, you mean that this stuff with ALL_SUBDIRS.  Presumably this is to help clean existing builds and would go away eventually?  That doesn't seem too important to me, but if you're going to do it for the runtime, I don't see why there isn't a transitional part in the compiler directory, which has the same problems.

**Comment by mppf** [comment]
@sungeunchoi - the transitional part will remove gen directories in the runtime for now on clobber/cleanall. I could remove that from this PR if you prefer. If it stays, I don't think it's worth making the corresponding changes to the compiler since it is more typical that only one compiler configuration is built. In particular, the compiler doesn't use the pattern of say runtime/src/launch. In that pattern there are many possible subdirs that could be used (depending on configuration) but cleanall should clean all of them. (make clean already cleans up the configured subdir).

But the real question here - do you think it would be better to remove the ALL_SUBDIRS stuff in this PR?

**Comment by sungeunchoi** [comment]
@mppf Yes, I think it's be better to omit the ALL_SUBDIRS stuff.

**Comment by mppf** [comment]
@sungeunchoi - OK, I reverted that part. Are we good to go now?

File: runtime/src/comm/gasnet/Makefile.share [hunk 1]
```diff
@@ -34,11 +34,13 @@ COMM_LAUNCHER_OBJS = \
 ifneq ($(MAKE_LAUNCHER), 1)
 ifneq ($(CHPL_MAKE_RULE),clean)
 ifneq ($(CHPL_MAKE_RULE),clobber)
+ifeq ($(CHPL_MAKE_COMM),gasnet)
```
Comment (sungeunchoi) [inline runtime/src/comm/gasnet/Makefile.share]: Maybe I'm misunderstanding something, but why is this extra guard desirable?

Comment (mppf) [inline runtime/src/comm/gasnet/Makefile.share]: It mattered with other changes but no longer does. I'll revert it.

**Comment by sungeunchoi** [comment]
👍

**Pull request merged**